sensor/debugger/k8s/fake.go

```diff
@@ -0,0 +1,296 @@
+		deploymentKind:  func(namespace string) interface{} { return f.Client.Kubernetes().AppsV1().Deployments(namespace) },
+		statefulSetKind: func(namespace string) interface{} { return f.Client.Kubernetes().AppsV1().StatefulSets(namespace) },
+		cronJobKind:     func(namespace string) interface{} { return f.Client.Kubernetes().BatchV1().CronJobs(namespace) },
+		podKind:         func(namespace string) interface{} { return f.Client.Kubernetes().CoreV1().Pods(namespace) },
+	}
+	f.resourceMap = map[string]interface{}{
+		namespaceKind:             &corev1.Namespace{},
+		secretKind:                &corev1.Secret{},
+		serviceAccountsKind:       &corev1.ServiceAccount{},
+		roleKind:                  &rbacv1.Role{},
+		clusterRoleKind:           &rbacv1.ClusterRole{},
+		roleBindingKind:           &rbacv1.RoleBinding{},
+		clusterRoleBindingKind:    &rbacv1.ClusterRoleBinding{},
+		networkPolicyKind:         &networkingv1.NetworkPolicy{},
+		nodeKind:                  &corev1.Node{},
+		serviceKind:               &corev1.Service{},
+		jobKind:                   &batchv1.Job{},
+		replicaSetKind:            &appsv1.ReplicaSet{},
+		replicationControllerKind: &corev1.ReplicationController{},
+		daemonSetKind:             &appsv1.DaemonSet{},
+		deploymentKind:            &appsv1.Deployment{},
+		statefulSetKind:           &appsv1.StatefulSet{},
+		cronJobKind:               &batchv1.CronJob{},
+		podKind:                   &corev1.Pod{},
+	}
+}
+
+// WaitForDone waits until all the events have been processed
+func (f *FakeEventsManager) WaitForDone() {
+	doneC := f.done.WaitC()
+	<-doneC
+}
+
+// waitForMinimumResources waits for a minimum number of resources to be created or once all the events have been processed
+func waitForMinimumResources(ch <-chan string, done concurrency.Signal) error {
+	count := 0
+	doneC := done.WaitC()
+	d := false
+	for {
+		select {
+		case obj, more := <-ch:
+			if !more && d {
+				return errors.New("the events file did not contain the minimum resources required to start sensor")
+			}
+			if _, ok := minimumResources[obj]; ok {
+				minimumResources[obj]--
+				if minimumResources[obj] == 0 {
+					count++
+				}
+				if len(minimumResources) == count {
+					return nil
+				}
+			}
+		case <-doneC:
+			// We make sure there are not more messages in the channel
+			obj, more := <-ch
+			if !more {
+				return errors.New("the events file did not contain the minimum resources required to start sensor")
+			}
+			if _, ok := minimumResources[obj]; ok {
+				minimumResources[obj]--
+				if minimumResources[obj] == 0 {
+					count++
+				}
+				if len(minimumResources) == 0 {
+					return nil
+				}
+			}
+			d = true
+		}
+	}
+}
+
+// CreateEvents creates the k8s events from a given jsonl file
+func (f *FakeEventsManager) CreateEvents() error {
+	f.done = concurrency.NewSignal()
+	objs, err := f.Reader.ReadFile()
+	ch := make(chan string, len(objs))
```

Comment:
The `length` of a channel isn't the same thing as the length of a slice. What you're defining here is a `buffered` channel that allows for sending data into a channel in a non-blocking way up until the buffer size. See [this as an example](https://go.dev/tour/concurrency/3).

I think it's fine to have a buffered channel and even more optimal here, but you have to consider this when handling the `done` signal. See my next comment.

Reply:
Yeah, it was a work in progress. After my chat with @vikin91 we changed this part a lot.